documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -11,203 +11,683 @@ Proposal API for backend-agnostic, refactored Input via keyboards, mice, and con
+
+```cs
+public static class InputWindowExtensions
+{
+    public static IInputBackend CreateInputBackend(this WindowHandles window);
+    public static IInputBackend CreateInputBackend(this IWindowHandlesSource window);
+    public static InputContext CreateInput(this WindowHandles window);
+    public static InputContext CreateInput(this IWindowHandlesSource window);
+}
+```
+
+The `CreateInputBackend` will create an instance of the reference implementation for the given `WindowHandles`. The `IWindowHandlesSource` overloads just forward to the `WindowHandles` overload. This is because `ISurface` will implement `IWindowHandlesSource`, so the extension methods will be usable on an `ISurface` without having a hard reference between Windowing and Input.
+
+The `CreateInput` methods simply return an `InputContext` preconfigured with the backend created by `CreateInputBackend` for ease of use.
+
+Please see the Windowing 3.0 proposal for `IWindowHandlesSource` and `WindowHandles`.
+
+# Devices, Backends, and Contexts
+
+Input devices all inherit from a root interface.
+
+```cs
+public interface IInputDevice
+{
+    nint BackendId { get; }
+    string Name { get; }
+}
+```
+
+`BackendId` is an integral identifier for this device unique to this backend **only** - i.e. it is not globally unique, and if you have an input context with multiple backends there may be multiple devices with a particular `BackendId`. It may be a handle to memory, or it may be a generic integral index - this is implementation-defined.
+
+`Name` is a rough description of the input device. Its value is not intrinsically meaningful.
+
+All devices originate from a context.
+
+```cs
+public interface IInputBackend
+{
+    string Name { get; }
+    nint Handle { get; }
+    IReadOnlyList<IInputDevice> Devices { get; }
+    void Update();
+}
+```
+
+`Name` is a rough description of the input backend. Its value is not intrinsically meaningful.
+
+`Handle` is a globally-unique handle to memory for this input backend. The memory this handle actually points to is implementation-defined.
+
+`Devices` enumerates all of the **connected** devices available from this input backend. When a device is disconnected, its `IInputDevice` object should be discarded by all that consumed it, as it can not be relied upon for being reused by the input backend. An implementation is welcome to reuse old objects, but this is strictly implementation-defined. A device not being present in the `Devices` list is sufficient evidence that a device has been disconnected.
+
+`Update` will update the state of all devices contained within this input backend. The value of the `State` properties on each device must not change until this method is called. This is a departure from 1.0's and 2.0's model of updating state as soon as new information is available, which has resulted in lots of inconsistencies in the past.
```

Comment:
Given that we now have a state model multi-threading may make more sense now.
So I feel this method is a little too relaxed. How exactly does it update the state? surely not atomically, that'd require a lock here and everywhere a state is read.
Why don't we just leave the updating of necessary devices to the user? I mean they can decide to, for example, update all the keyboards.
I think instead we should just provide a method that updates the connection status of devices, so updates the `Devices` list (atomically). From there users are free to enumerate the devices themselves and update as necessary.

Reply:
No user will care to do this, and asking the user to update each device or device list themselves is much more effort than the average user will care to put in.

On threading, I have updated the proposal to explicitly forbid using the `Update` method anywhere but the main thread as this is illegal on a couple of operating systems and by extension is illegal in the underlying APIs.

```suggestion
`Update` will update the state of all devices contained within this input backend. The value of the `State` properties on each device must not change until this method is called. This is a departure from 1.0's and 2.0's model of updating state as soon as new information is available, which has resulted in lots of inconsistencies in the past.

The onus is on the user to coordinate using this type across threads, as the input context is not thread safe. In addition, certain backends may have (unavoidable) restrictions on what thread `IInputBackend.Update` can be called on - the user is responsible for respecting these threading rules as well.

Threading rules for the reference implementation (if any) will be explicitly documented, and guidance for using this type safely against any first-party implementations will be included in the XML and user documentation.

**INFORMATIVE TEXT:** For example, it is illegal for GLFW functions to be called anywhere except the thread `glfwInit` was called on, and it is illegal on some operating systems (such as macOS) for `glfwInit` to be called anywhere except the thread that called `main`.
```

Reply:
This doesn't resolve the conflict.
Does the proposal suggest that calling backed / context `.Update` atomically updates all devices? I wouldn't think so, that'd require a lock for every state read.
In that case we may have cases where a background thread reads states (idk, enumerates devices and looks at each of them) and while it does this states are being updated, making the cross-device view inconsistent.

Reply:
> Does the proposal suggest that calling backed / context .Update atomically updates all devices? I wouldn't think so, that'd require a lock for every state read.

Nope.

> In that case we may have cases where a background thread reads states (idk, enumerates devices and looks at each of them) and while it does this states are being updated, making the cross-device view inconsistent.

Yep.

> (I'm not suggesting we can't just say "this is by design, users have to deal with this" but I'd want a specific note that this is a potential issue)

Does my new comment not cover this? Namely `The onus is on the user to coordinate using this type across threads, as the input context is not thread safe.`. If not, can you suggest an elaborated sentence?